drivers: video: ov2640: add sensor driver for ov2640

robertszczepanski opened the pull request:
This adds driver for Omnivision OV2640 image sensor.

The driver provides support for:
* Camera initialization.
* 10 camera controls - there is support for old control macros like horizontal and vertical flip, exposure and gain control. There are also new control macros that support brightness, saturation, white balance, contrast, colorbar and quality settings.
* Camera hardware reset.
* 2 output formats - RGB565 and JPEG.
* 10 resolutions in both supported formats (160x120, 176x144, 240x160, 320x240, 352x288, 640x480, 800x600, 1024x768, 1280x1024, 1600x1200).

There is also new node in `litex_vexriscv` devicetree that represents OV2640 camera.
It has been tested on [Arty A7-35T](https://reference.digilentinc.com/programmable-logic/arty-a7/start) with [LiteX SoC](https://github.com/enjoy-digital/litex). [OV2640 camera](https://www.waveshare.com/ov2640-camera-board.htm) XCLK set to 24MHz.

loicpoulain commented on drivers/video/Kconfig.ov2640:
```diff
@@ -0,0 +1,10 @@
+# OV2640
+
+# Copyright (c) 2021 Antmicro <www.antmicro.com>
+# SPDX-License-Identifier: Apache-2.0
```
The drivers/video/ov2640_regs.h file  has a different license (MIT), could you clarify this?

robertszczepanski replied:
I've done some changes to use Apache 2.0 license.

loicpoulain commented on include/drivers/video-controls.h:
```diff
@@ -44,6 +44,12 @@ extern "C" {
+#define VIDEO_CID_CAMERA_CONTRAST	(VIDEO_CTRL_CLASS_CAMERA + 6)
+#define VIDEO_CID_CAMERA_COLORBAR	(VIDEO_CTRL_CLASS_CAMERA + 7)
+#define VIDEO_CID_CAMERA_QUALITY	(VIDEO_CTRL_CLASS_CAMERA + 8)
 
```
It looks good, but could you please move video generic/framework change to a dedicated patch.

robertszczepanski replied:
Changes in `video.h` and `video-controls.h` are now in separate commits.

loicpoulain commented:
> This adds driver for Omnivision OV2640 image sensor.
> 
> The driver provides support for:
> 
>     * Camera initialization.
> 
>     * 10 camera controls - there is support for old control macros like horizontal and vertical flip, exposure and gain control. There are also new control macros that support brightness, saturation, white balance, contrast, colorbar and quality settings.
> 
>     * Camera hardware reset.
> 
>     * 2 output formats - RGB565 and JPEG.

> There is also new node in `litex_vexriscv` devicetree that represents OV2640 camera.
> It has been tested on [Arty A7-35T](https://reference.digilentinc.com/programmable-logic/arty-a7/start) with [LiteX SoC](https://github.com/enjoy-digital/litex). [OV2640 camera](https://www.waveshare.com/ov2640-camera-board.htm) XCLK set to 24MHz.

Do you plan to submit a subsequent patch for the image processing/acquisition side of this microcontroller?

robertszczepanski committed:
drivers: video: add JPEG pixel format

Add JPEG pixel format needed for cameras with JPEG
compression capability.

Signed-off-by: Robert Szczepanski <rszczepanski@antmicro.com>

robertszczepanski committed:
drivers: video: add more camera control IDs

Add more camera control IDs for cameras with more capabilities.

Signed-off-by: Robert Szczepanski <rszczepanski@antmicro.com>

robertszczepanski committed:
drivers: video: ov2640: add driver for ov2640 image sensor

This adds the driver for Omnivision OV2640 image sensor.

The driver provides support for 10 different resolutions in range from
160x120 to 1600x1200 in both JPEG and RGB565 pixel formats. There are
also mutliple configuration options, e.g. hflip, vflip, saturation and
brightness control.

Signed-off-by: Robert Szczepanski <rszczepanski@antmicro.com>

robertszczepanski commented:
Right now we don't have any subsequent patches for the image processing planned for `litex_vexriscv`.

loicpoulain commented on drivers/video/ov2640_regs.h:
```diff
@@ -0,0 +1,410 @@
+/*
```
AFAIU, you can put all this content in .c file directly, no gain in having a .h if nothing is shared or exported.

robertszczepanski replied:
Right, merged them.

loicpoulain commented:
> Right now we don't have any subsequent patches for the image processing planned for `litex_vexriscv`.

ok, understood, but how did you test that? have you only tested the control path (i2c) or are you using some downstream driver for data catpure?

robertszczepanski commented:
I developed an app that integrates this driver with [FastVDMA](https://github.com/antmicro/fastvdma), Zephyr net driver and shell subsystem. This allows me to write captured image to memory and then transfer it to my PC via Ethernet. After that I am able convert an image to png (using `ffmpeg`) and display it on screen.
Shell subsystem with additional commands works great here for camera testing. I am able to get multiple captures (one at the time) with different camera settings without resetting device.

loicpoulain commented:
@ro

> I developed an app that integrates this driver with [FastVDMA](https://github.com/antmicro/fastvdma), Zephyr net driver and shell subsystem. This allows me to write captured image to memory and then transfer it to my PC via Ethernet. After that I am able to convert an image to png (using `ffmpeg`) and display it on screen.
> Shell subsystem with additional commands works great here for camera testing. I am able to get multiple captures (one at the time) with different camera settings without resetting device.

Ok, thanks for the clarification.

loicpoulain commented on boards/riscv/litex_vexriscv/litex_vexriscv.dts:
```diff
@@ -51,6 +51,13 @@
+		compatible = "ovti,ov2640";
+		reg = <0x30>;
+		label = "OV2640";
+		status = "okay";
```
Is this sensor always connected to that board? is it an 'official' expension board? or a custom setup?

robertszczepanski replied:
It's custom setup, I used [this](https://www.waveshare.com/ov2640-camera-board.htm) camera as mentioned before.

loicpoulain replied:
Ok, so you can not enable this specific device in the mainline dts, since it's specific to your setup. You can either remove that change from the series and keep your change downstream, or alternatively I think you can add the node but with `status = "disabled"` by default with proper documentation to explain how replicating your setup. First solution is probably simpler ;-)

robertszczepanski replied:
Thanks, I guess I'll get rid of that part then :)
I removed that commit.

loicpoulain commented on drivers/video/ov2640.c:
```diff
@@ -0,0 +1,1057 @@
+	int ret = 0;
+	int i = 0;
+
+	/* we only support one format for now (VGA RGB565) */
```
and JPEG...

robertszczepanski replied:
Sure, changed that.

loicpoulain commented:
LGTM from a video framework perspective.

carlescufi commented:
@robertszczepanski please rebase to the latest main branch

robertszczepanski commented:
@carlescufi done :)
